ambry-api/src/main/java/com.github.ambry/rest/RequestPath.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.rest;
+
+import java.util.List;
+
+
+public class RequestPath {
+  private final String prefix;
+  private final String clusterName;
+  private final String operationOrBlobId;
+  private final RestUtils.SubResource subResource;
+  private String operationOrBlobIdWithoutLeadingSlash = null;
+
+  /**
+   * Parse the request path (and additional headers in some cases). The path will match the following regex-like
+   * description:
+   * <pre>
+   * {@code {prefixToRemove}?{clusterNameSegment}?({operationOrBlobId}|{operationOrBlobId}/{subResource})}
+   * </pre>
+   * For example, something like {@code /prefixOne/clusterA/parts/of/blobId/BlobInfo}, {@code /clusterA/operation},
+   * or {@code /blobId}, among others.
+   * @param restRequest {@link RestRequest} containing metadata about the request.
+   * @param prefixesToRemove the list of prefixes that could precede the other parts of the URL. Removal of
+   *                         prefixes earlier in the list will be preferred to removal of the ones later in the list.
+   * @param clusterName the cluster name to recognize and handle when parsing the URL. Case is ignored when matching
+   *                    this path segment.
+   * @return a {@link RequestPath} object.
+   */
+  public static RequestPath parse(RestRequest restRequest, List<String> prefixesToRemove, String clusterName) {
```

Comment:
correct me if I am wrong,  I didn't see the query parameters (if there are any in the `restRequest`) are parsed from the uri, are they extracted somewhere else?

Reply:
The removal of the query parameters in this method wasn't actually needed since the `RestRequest` interface requires that `getPath()` will exclude any query parameters:
```java
  /**
   * Return the path (the parts of the URI after the domain excluding query parameters).
   * @return path String that represents part of the uri excluding domain and query parameters.
   */
  String getPath();
```
The `RestRequest` implementation should parse the query parameters during construction and include them in the map returned by `getArgs()`